lib/lua/TJsonProtocol.lua

```diff
@@ -0,0 +1,725 @@
+end
+
+-- decoding
+function base64_decode(data)
+    data = string.gsub(data, '[^'..b..'=]', '')
+    return (data:gsub('.', function(x)
+        if (x == '=') then return '' end
+        local r,f='',(b:find(x)-1)
+        for i=6,1,-1 do r=r..(f%2^i-f%2^(i-1)>0 and '1' or '0') end
+        return r;
+    end):gsub('%d%d%d?%d?%d?%d?%d?%d?', function(x)
+        if (#x ~= 8) then return '' end
+        local c=0
+        for i=1,8 do c=c+(x:sub(i,i)=='1' and 2^(8-i) or 0) end
+        return string.char(c)
+    end))
+end
+
+function TJSONProtocol:resetContext()
+  self.jsonContext = {}
+  self.jsonContextVal = {first = true, colon = true, ttype = 2, null = true}
+  self.jsonContextIndex = 1
+end
+
+function TJSONProtocol:contextPush(context)
+  self.jsonContextIndex = self.jsonContextIndex + 1
+  self.jsonContext[self.jsonContextIndex] = self.jsonContextVal
+  self.jsonContextVal = context
+end
+
+function TJSONProtocol:contextPop()
+  self.jsonContextVal = self.jsonContext[self.jsonContextIndex]
+  self.jsonContextIndex = self.jsonContextIndex - 1
+end
+
+function TJSONProtocol:escapeNum()
+  if self.jsonContextVal.ttype == 1 then
+    return self.jsonContextVal.colon
+  else
+    return false
+  end
+end
+
+function TJSONProtocol:writeElemSeparator()
+  if self.jsonContextVal.null then
+    return
+  end
+  if self.jsonContextVal.first then
+    self.jsonContextVal.first = false
+  else
+    if self.jsonContextVal.ttype == 1 then
+      if self.jsonContextVal.colon then
+        self.trans:write(JSONNode.PairSeparator)
+        self.jsonContextVal.colon = false
+      else
+        self.trans:write(JSONNode.ElemSeparator)
+        self.jsonContextVal.colon = true
+      end
+    else
+      self.trans:write(JSONNode.ElemSeparator)
+    end
+  end
+end
+
+function TJSONProtocol:hexChar(val)
+  val = libluabitwise.band(val, 0x0f)
+  if val < 10 then
+    return val + 48
+  else
+    return val + 87
+  end
+end
+
+function TJSONProtocol:writeJSONEscapeChar(ch)
+  self.trans:write(JSONNode.EscapePrefix)
+  local outCh = hexChar(libluabitwise.shiftr(ch, 4))
+  local buff = libluabpack.bpack('c', outCh)
+  self.trans:write(buff)
+  outCh = hexChar(ch)
+  buff = libluabpack.bpack('c', outCh)
+  self.trans:write(buff)
+end
+
+function TJSONProtocol:writeJSONChar(byte)
+  ch = string.byte(byte)
+  if ch >= 0x30 then
+    if ch == JSONNode.Backslash then
+      self.trans:write(JSONNode.Backslash)
+      self.trans:write(JSONNode.Backslash)
+    else
+      self.trans:write(byte)
+    end
+  else
+    local outCh = JSONCharTable[ch+1]
+    if outCh == 1 then
+      self.trans:write(byte)
+    elseif outCh > 1 then
+      self.trans:write(JSONNode.Backslash)
+      local buff = libluabpack.bpack('c', outCh)
+      self.trans:write(buff)
+    else
+      self:writeJSONEscapeChar(ch)
+    end
+  end
+end
+
+function TJSONProtocol:writeJSONString(str)
+  self:writeElemSeparator()
+  self.trans:write(JSONNode.StringDelimiter)
+  -- TODO 
```

Comment:
It can be escaped by being preceded with a backslash,
like " to \", \t to \\t, \r to \\r, \f to \\f, \n to \\n , \b to \\b.
Otherwise, character " will break JSON format.
e.g:
To be sent string is "Begin:This is Send\"String\b\t\n\r\f:End."